test/util/TestLockFile.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "../../source/util/LockFile.h"
+#include "gtest/gtest.h"
+
+#include <fstream>
+#include <thread>
+#include <unistd.h>
+
+using namespace std;
+using namespace Aws::Iot::DeviceClient::Util;
+
+TEST(LockFile, normalCreation)
+{
+    string path = "/run/lock/";
+    unique_ptr<LockFile> lockFile = unique_ptr<LockFile>(new LockFile{path, "./aws-iot-device-client"});
+
+    ifstream fileIn(path + "devicecl.lock");
+    ASSERT_TRUE(fileIn);
+
+    string storedPid;
+    if (fileIn >> storedPid)
+    {
+        ASSERT_STREQ(to_string(getpid()).c_str(), storedPid.c_str());
+    }
+}
+
+TEST(LockFile, earlyDeletion)
+{
+    string path = "/run/lock/";
+    unique_ptr<LockFile> lockFile = unique_ptr<LockFile>(new LockFile{path, "test-aws-iot-device-client"});
+    lockFile.reset();
+
+    ifstream fileIn(path + "devicecl.lock");
+    ASSERT_FALSE(fileIn);
+}
+
+TEST(LockFile, multipleFiles)
+{
+    string path = "/run/lock/";
+    unique_ptr<LockFile> lockFile = unique_ptr<LockFile>(new LockFile{path, "test-aws-iot-device-client"});
+
+    EXPECT_THROW(unique_ptr<LockFile>(new LockFile{path, "test-aws-iot-device-client"}), std::runtime_error);
+}
+
+TEST(LockFile, multipleFilesWithExtendedPath)
+{
+    string path = "/run/lock/";
+    unique_ptr<LockFile> lockFile = unique_ptr<LockFile>(new LockFile{path, "test-aws-iot-device-client"});
+
+    EXPECT_THROW(unique_ptr<LockFile>(new LockFile{path, "directory/test-aws-iot-device-client"}), std::runtime_error);
+}
+
+TEST(LockFile, staleFile)
+{
+    string path = "/run/lock/";
+    string pidMax;
+    ifstream pidFile("/proc/sys/kernel/pid_max");
+    if (pidFile && pidFile >> pidMax)
+    {
+        ofstream fileOut(path);
```

Comment:
[bug] This line will open `/run/lock` for output. Since this is a directory, the call will fail and nothing is written. I think your intent was to create a file named `path + "devicecl.lock"`. Keep it DRY by creating a variable named `filename` and using that in both places in this branch.

Reply:
Actually, I changed the code to no longer require the full path including the filename.

You now only have to specify the parent directory and device client will just create a file named "devicecl.lock" in that dir.

I suppose this takes away the option for giving the lockfile a different name but I don't know if such a use case will ever come up.